akka-http-core/src/main/scala/akka/http/impl/engine/server/HttpServerBluePrint.scala

```diff
@@ -83,34 +84,38 @@ private[http] object HttpServerBluePrint {
   def controller(settings: ServerSettings, log: LoggingAdapter): BidiFlow[HttpResponse, ResponseRenderingContext, RequestOutput, RequestOutput, Unit] =
     BidiFlow.fromGraph(new ControllerStage(settings, log)).reversed
 
-  def requestPreparation(settings: ServerSettings, remoteAddress: Option[InetSocketAddress])(implicit mat: Materializer): BidiFlow[HttpResponse, HttpResponse, RequestOutput, HttpRequest, Unit] = {
-    import settings._
+  def requestPreparation(settings: ServerSettings)(implicit mat: Materializer): BidiFlow[HttpResponse, HttpResponse, RequestOutput, HttpRequest, Promise[Option[Option[InetSocketAddress]]]] = {
+    // yes that's the type we actually need... so much can go wrong here, proper completion is (Some(Some(addr)), otherwise one has to Some(None)...
+    // becaue of completion semantics of success(None) on Source.maybe
+    val remoteAddressSource: Source[Option[InetSocketAddress], Promise[Option[Option[InetSocketAddress]]]] =
+      Source.maybe[Option[InetSocketAddress]]
 
     val prepareRequest =
       Flow[RequestOutput]
         .splitWhen(x ⇒ x.isInstanceOf[MessageStart] || x == MessageEnd)
         .via(headAndTailFlow)
+        .zipWithMat(remoteAddressSource)({ case ((o, src), addr) ⇒ (o, src, addr) })(Keep.right)
```

Comment:
I'd probably add the Option[InetAddress] to RequestStart instead of creating yet another tuple.
.zipWithMat(remoteAddressSource)((hAt, addr) => (hAt._1.withRemoteAddress(addr), hAt._2))(Keep.right)